File: 0834-sum-of-distances-in-tree/0834-sum-of-distances-in-tree.py
from collections import defaultdict
from typing import List

class Solution:
    def sumOfDistancesInTree(self, n: int, edges: List[List[int]]) -> List[int]:
        graph = defaultdict(list)
        for u, v in edges:
            graph[u].append(v)
            graph[v].append(u)
        count = [1] * n
        ans = [0] * n 
        def dfs_bottom_up(curr_node: int, parent_node: int) -> None:
            for child in graph[curr_node]:
                if child == parent_node:
                    continue
                dfs_bottom_up(child, curr_node)
                count[curr_node] += count[child]
                ans[curr_node] += ans[child] + count[child]
        def dfs_top_down(curr_node: int, parent_node: int) -> None:
            for child in graph[curr_node]:
                if child == parent_node:
                    continue
                ans[child] = ans[curr_node] - count[child] + (n - count[child])
                dfs_top_down(child, curr_node)
        dfs_bottom_up(0, -1)
        dfs_top_down(0, -1)
        
        return ans

File: 0834-sum-of-distances-in-tree/0834-sum-of-distances-in-tree.cpp
class Solution {
public:
    vector<int> sumOfDistancesInTree(int n, vector<vector<int>>& edges) {
        // int n = edges.size();
        vector<vector<int>>tree(n);
        for(auto itr: edges){
            tree[itr[0]].push_back(itr[1]);
            tree[itr[1]].push_back(itr[0]);
        }
        vector<int>nodecount(n);
        vector<int>visited(n);
        vector<int>distance(n);
        long long root_result = 0;
        auto DFS = [&](auto &&self,int node,int parent,int curr_depth)->long long{
            long long count = 1;
            root_result += curr_depth;
            for(auto &negh:tree[node]){
                if(negh == parent){
                    continue;
                }
                count += self(self,negh,node,curr_depth+1);
            }

            return nodecount[node] = count; 
        };
        DFS(DFS,0,-1,0);
        distance[0] = root_result;
        auto DFS2 = [&](auto&&self,int node,int parent) ->void{
            for(int &child :tree[node]){
                if(child==parent)continue;
                distance[child] = distance[node] - 2*nodecount[child] + n;
                self(self,child,node);
            }
        };
        DFS2(DFS2,0,-1);
        return distance;
    }
};